modules/concurrency/src/spinner.cpp

```diff
@@ -0,0 +1,56 @@
+//=================================================================================================
+// Copyright (C) 2023-2024 HEPHAESTUS Contributors
+//=================================================================================================
+
+#include "hephaestus/concurrency/spinner.h"
+
+#include <fmt/format.h>
+
+#include "hephaestus/base/exception.h"
+
+namespace heph::concurrency {
+Spinner::Spinner() : is_started_(false) {
+}
+
+Spinner::~Spinner() {
+  try {
+    if (is_started_.load() && spinner_thread_.joinable()) {
+      stop();
+    }
+  } catch (const std::exception& ex) {
+    fmt::print("Error stopping spinner: {}", ex.what());
+  }
+}
+
+void Spinner::start() {
+  throwExceptionIf<InvalidOperationException>(is_started_.load(), fmt::format("Spinner is already started."));
+
+  spinner_thread_ = std::jthread([this](std::stop_token stop_token) { spin(stop_token); });
+
+  is_started_.store(true);
+}
+
+void Spinner::spin(std::stop_token& stop_token) {
+  while (!stop_token.stop_requested()) {
+    spinOnce();
```

Comment:
Do you remember in arrival that we had issue of exception not being reported properly, was that because of std::async right? here we should be good I think

Reply:
yes, I think it was because of the async.